### ISSUE TITLE
Update Data Table

### ISSUE THREAD

**Issue by xNS5** [xNS5/rentalreviews#47]
- [x] Change link icons to something that better communicates the browser will redirect to something in the same tab
- [x] Update data table component to indicate which sorting is being used, and the direction that it's currently being sorted. Presently it's not clear which is being used.
  - Ideas: Hide/Show the up/down arrow based on current sort direction, and show both if not currently being used.

**Comment by xNS5** [comment]
Instead of simply performing a hide/show when the sorting is being used, I created a `switch(...){...}` statement with 3 conditions: first, "asc" which not only returns an ascending sort icon but also returns the relevant aria-label. Second "desc" which does the same as "asc" but with a "descending" aria-label text and icon. Lastly, a "disabled" aria label and icon to indicate that the column isn't currently being sorted.

**Comment by xNS5** [comment]
This is a good link to keep on hand just in case: https://tink.uk/how-screen-readers-navigate-data-tables/

**Comment by xNS5** [comment]
I'm honestly pretty frustrated with the shadcn/tanstack table. I'm not sure if it's the component itself or my own inexperience, but I'm unable to navigate the table using my screen reader. It reads out the row, which I don't know if that's *supposed* to happen, but traversing the table itself is challenging. I could add tab index values but that would lock the user into tabbing a bunch until they reach the end of the list.

**Comment by xNS5** [comment]
I created a new branch 47-react-aria as a sort of experiment to see what needs to be done to make my current table more accessible. After all, it's called "react-aria" -- what could go wrong?

**Comment by xNS5** [comment]
React-aria doesn't have any kind of pagination, so I took some notes on the behaviors of the react-aria table component:
* All cells and rows (including the header) has `tabIndex="-1"` -- which means the element can be focused but isn't in the main tabindex order.
* I use the arrow keys for navigating rows as well as cells.
* Seems like the focus behavior is row -> cell. If the row is selected, then it navigates up and down. If I hit an arrow key while the row is selected, it goes to that row's first cell. 

I'm going to experiment a bit on how to implement cell navigation (up/down/right/left) with the screen reader.

**Comment by xNS5** [comment]
I've implemented some basic screen reader navigation that assumes the user will be using `ctrl + alt + arrow` for traversing the table cells with the following logic:

1. If the user is on the header row and navigates down, the focus will be on the cell in the same column in the first row of the table.
2. If the user is at the top of the table and navigates up, the focus will be on the header row cell in the same column.
3. If the user hits "enter" and the element is a button, it will click on that button.
  - The flaw here is that if the user strictly uses tab-based navigation, if they hit enter after using the `ctrl + alt` combo the button will only click once -- which is a problem.
4. If the user is in the table and does `ctrl + alt + (up | down | right | left)` the focus will shift in that direction until they hit the end of the column/row, or beginning of the column/row.


I need to figure out how to handle strictly tab-based interaction now and switch back and forth between them. Ugh.

**Comment by xNS5** [comment]
Another possibility: it could be an issue with just ORCA. Will test on a Windows machine.

**Comment by xNS5** [comment]
Tested using NVDA and *man* is the tanstack table finnicky. I think it might be better for me to use `react-aria` table instead -- just eyeballing it, it looks a lot simpler and I'm not having the kinds of issues I'm having with the tanstack table.

**Comment by xNS5** [comment]
Note to self: update the search feature with a search hint to indicate that the search bar is to filter company names.

**Comment by xNS5** [comment]
I've implemented a basic data table using React-Aria's table component. Overall code structure mirrors shadcn code for consistency with the base component functionality in `ui/aria-table.tsx` and component-specific logic in `aria-table/data-table.tsx`. 

Overall it's a smidge more confusing than the React table, but I have high hopes for data accessibility with this table. Next steps will be to test the table with NVDA and Orca, then integrate pagination and a search feature if the evaluation passes my standards.

**Comment by xNS5** [comment]
Did some testing using Orca, and the results look very promising.

Added the following functionality:
- Pagination
- Column 0 is a link to the review page
- Cell 0 in Row **n** has onHover to display link icon
- Navigation buttons that conditionally enable and disable based on the current page index
- `onClick` handlers for buttons
- `onMouseEnter` and `onMouseLeave` handlers for link icon
- Sorted data, and paginated data based on the sorted data. 
   - In essence, all of the data gets sorted based on the column which is by default the "name" column. After that, the `currentPageData` is the paginated data. 
- Page change handler which sets the current page to a number **n** that's between 1 and `data.length`. Also added a smidge more logic in the link hover state which clears the states when moving to a new page. It doesn't make much sense for that data to persist. 


TODO:

- [x] Add search functionality
- [x] Add `aria-label` where applicable from the `alt` object
- [x] Test with NVDA on Windows

**Comment by xNS5** [comment]
In order to implement search functionality, I needed to re-organize how the data table is populated.

First, created a state for the input field value. Paired with an `onChange(...)` event, the input field changes and keeps the value. 

Made some small tweaks to the sorted data:

* First, if `searchTerm.length > 0`, that means the user is using the search bar (intentionally or otherwise). I used the `useMemo(....)` hook to filter the data based on the "name" column. Felt easier that way, for now at least. 
* Then, the data from `filteredData` will be sorted based on the filtered data. 
* Lastly, the data is paginated.

Misc additions:
* I created a memoized `pageCount` value which calculates the maximum number of pages using `Math.ceil(sortedData.length / paginationValue)` where `paginationValue` is passed in via props or is the default value of 10. 

Notes:

* Testing out using ORCA seems to work fine, will have to do some more tests
* Need to test using NVDA.

**Comment by xNS5** [comment]
Tested using NVDA, I needed to add some focus rings to display when the screen reader is focused on a cell.

**Comment by xNS5** [comment]
Forgot to mention this, but I also experienced some odd behavior with NVDA. I needed to tab to a given element on the table before I was able to use `ctrl + alt + (up|down|right|left)` to navigate the table. The fact that I was able to period demonstrates that React-Aria is going to suit my needs better, but it was an oddity. 

I'll review the requirements for updates to this component in the AM, and assuming everything is to spec I'll merge with development.
